### PR TITLE
Add menu item reference search across all item types

### PR DESCRIPTION
**Pull request by jimcottrell**
This pull request may require some review/discussion.  The static menu item type system provides a lot of development flexibility, but particularly for clients we thought it would be valuable to have a search option that would work across item types. Since a site may make use of multiple custom content types, sometimes it might get a little confusing. The attached GIF shows this pull request's functionality.

The other behavioral change made is that selecting a reference will update the title field if the last-saved title is still the default new item title. This happens both with the search and with manual reference selection, and is more clearly shown in the second menu item edited in the GIF. Since there's no "raw" title, there's a bit of a hack in this to remove the added CMS page base file name suffix, which maybe isn't the right way to go.
        
I hadn't dug much into October widgets before this, so I struggled a bit with the "right" way to route AJAX requests to the reference search, given that the form containing the widget isn't normally instantiated and is nested within other content. If there's a better way, I'd love to know about it. And of course I'm open to any other changes that need to be made before this can be accepted.

demo

### CODE REVIEW

**Comment by LukeTowers** [comment]
Wow, this looks pretty impressive from a functionality standpoint! I don't have a ton of time to dig into your code right now, but one thing I would mention about the interface is: 

Are you able to make the "Search References" widget more closely resemble the type dropdown with the search bar integrated directly into the same dropdown that results display in? That would make the UX a bit more consistent with the rest of the dropdowns within October.

I'm definitely interested in getting this merged, we'll just need to get the specifics nailed down a bit first :)

**Comment by jimcottrell** [comment]
That's true; the UX is kind of its own thing.  I tried to limit the new code required, but I think to get a more consistent feel there I'd have to interface with Select2 directly, unless there's some similar functionality somewhere that I've overlooked.  I'll look into what that might require.

**Comment by jimcottrell** [comment]
Having looked into this, leveraging the backend framework's AJAX request functionality actually makes this pretty simple. Adding an AJAX source seems like something that might be useful added to the core dropdown functionality (which is how I implemented it for this proof of concept).

menu-items-search

So, this works, but then I think the question is: is this the right thing to do or should I actually be trying to replace the type and reference dropdowns? The separate search field could maybe be a little confusing by providing two ways to accomplish the same task. A more complete replacement of the current input methods might also bring its own confusion though, and of course it would be more work. Thoughts?

If it's acceptable as is, I'll push changes here and submit a pull request to core.

**Comment by LukeTowers** [comment]
I think I like it as it is, the search box being separate from the type and reference boxes. The reference box doesn't exist for every type, so it makes sense to keep them separate. Question, what's the Please enter 2 or more characters message about?

**Comment by jimcottrell** [comment]
It's part of the underlying Select2 UI.  I used the minimumInputLength option to prevent any requests from taking place until there are at least two characters to search on.  Select2 supports internationalization, of course, but I don't know if October makes use of that or if we'd want to not use that option.

**Comment by LukeTowers** [comment]
I don't think that option is used on any other dropdowns in october (minimumInputLength)

**Comment by jimcottrell** [comment]
I'm sure you're right.  I think we need to either use that option or otherwise limit search results though, since pulling all references across all types (i.e. an empty search) probably isn't the right thing to do.

**Comment by LukeTowers** [comment]
Let's get @daftspunk's opinion on the matter

**Comment by LukeTowers** [comment]
Valid logic. I'll pester @daftspunk to take a look at this later today. You mind pushing your most recent changes to this PR?

**Comment by jimcottrell** [comment]
@LukeTowers Do you want a separate pull request for the custom select AJAX functionality in core?  Or do you just want me to post the code change (it's not much) in a comment here?

File: formwidgets/menuitems/assets/js/menu-items-editor.js [hunk 1]
```diff
@@ -81,11 +81,19 @@
             self.$popupForm = self.$popupContainer.find('form')
             self.itemSaved = false
 
-            $('input[name=title]', self.$popupContainer).focus().select()
-            $('select[name=type]', self.$popupContainer).change(function(){
+            var titleField = $('input[name=title]', self.$popupContainer).focus().select()
+            var typeField = $('select[name=type]', self.$popupContainer).change(function(){
```
Comment (LukeTowers) [inline formwidgets/menuitems/assets/js/menu-items-editor.js]: minor quibble here, shouldn't variables containing jQuery objects be named prefixed with `$`? i.e. `$titleField` and `$typeField`?

**Comment by LukeTowers** [comment]
@jimcottrell Post the code that needed to change in core here and I'll let you know if we should have a separate PR for it

File: formwidgets/menuitems/assets/js/menu-items-editor.js (hunk 1 above)
Comment (jimcottrell) [inline formwidgets/menuitems/assets/js/menu-items-editor.js]: Yeah, pardon my inconsistency there.  I've corrected that.

**Comment by jimcottrell** [comment]
@LukeTowers See https://github.com/aspendigital/october/commit/376a3b59e2a7bf2738f8c26c8dfdb99df5f568e1

**Comment by LukeTowers** [comment]
So what do those changes do exactly?

**Comment by jimcottrell** [comment]
It adds [data-source] on a select element as an analog to [data-request] for AJAX functionality.  Select2 allows defining the transport used to pull data, so this passes the Select2 data access through $.fn.request to leverage all the other built-in functionality (setting headers, defining extra request data, etc.).

The changes to the formatting are there because AJAX data would be passed in for formatting without data attributes for the icon and image on an option element.

The possible issue is the internationalization as raised earlier, which I guess has always been an issue that plugin authors could expose if they made direct use of Select2 or set options via the data API.  Though likely no one ever has.

**Comment by LukeTowers** [comment]
I'm not hugely familiar with select2, but why does `data-source` need to be named `data-source`? Why can't it be named `data-request`? Just trying to understand the code here better for my own sake, not necessarily suggesting any changes.

**Comment by jimcottrell** [comment]
I used a different name so that the event listener in modules/system/assets/js/framework.js wouldn't trigger and send a request on change.  I suppose we could also prevent event propagation, but that was my first thought.

**Comment by LukeTowers** [comment]
On second thought, that makes the most sense to me as well. It's not really a generic use of the AJAX framework, so using `data-request` actually makes the effect more confusing. I like the split off of using `data-source`. Good choice.

Another question for you (sorry for all the questions 😄 ): Why implement the reference search as a Widget instead of a FormWidget?

**Comment by jimcottrell** [comment]
Yeah, that was my thinking on the naming.  Good to see it makes sense to you too. 😄 

I think the Widget/FormWidget thing has more to do with my inexperience, not having done anything with backend widgets before.  It probably would be more appropriate as a FormWidget.

**Comment by LukeTowers** [comment]
Plus you wouldn't have to add an AJAX handler to the controller directly if it's a FormWidget. 

If you've got any questions about implementing it as a FormWidget instead then let me know. Otherwise give it a go at making it a FormWidget instead :)

**Pull request merged**